cmake/add_cpp_test.cmake

```diff
@@ -1,38 +1,69 @@
 # Helper so we don't have to repeat ourselves so much
-# Usage: add_cpp_test(testname [COMPILE_ONLY] [SOURCES a.cpp b.cpp ...] [LABELS acceptance per_implementation ...])
-# SOURCES defaults to testname.cpp if not specified.
+# Usage: add_cpp_test(<testname>
+#                     [COMPILE_ONLY]
+#                     [DEBUG_ONLY]
+#                     [WILL_FAIL]
+#                     [LIBRARY]
+#                     [SOURCES a.cpp b.cpp ...]
+#                     [LABELS acceptance per_implementation ...]
+#                     [DEPENDENCY_OF ...])
+# SOURCES defaults to <testname>.cpp if not specified.
 function(add_cpp_test TEST_NAME)
   # Parse arguments
-  cmake_parse_arguments(PARSE_ARGV 1 ARGS "COMPILE_ONLY;LIBRARY;WILL_FAIL" "" "SOURCES;LABELS;DEPENDENCY_OF")
-  if (NOT ARGS_SOURCES)
+  cmake_parse_arguments(PARSE_ARGV 1 ARGS "COMPILE_ONLY;LIBRARY;WILL_FAIL;DEBUG_ONLY" "" "SOURCES;LABELS;DEPENDENCY_OF")
+
+  # Is generator multi-config?
+  get_cmake_property(is_multi_config GENERATOR_IS_MULTI_CONFIG)
+
+  # Return if debug only test
+  if(NOT is_multi_config AND ARGS_DEBUG_ONLY AND NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+    return()
+  endif()
+
+  # Set configurations
+  set(config_args)
+  if(is_multi_config AND ARGS_DEBUG_ONLY)
+    set(config_args CONFIGURATIONS Debug)
+  endif()
+
+  if(NOT ARGS_SOURCES)
     list(APPEND ARGS_SOURCES ${TEST_NAME}.cpp)
   endif()
-  if (ARGS_COMPILE_ONLY)
-    list(APPEND ${ARGS_LABELS} compile_only)
+
+  # Wrap sources in a generator expression and provide a dummy source
+  if(is_multi_config AND ARGS_DEBUG_ONLY)
+    list(TRANSFORM ARGS_SOURCES PREPEND "$<$<CONFIG:Debug>:")
+    list(TRANSFORM ARGS_SOURCES APPEND ">")
+    list(PREPEND ARGS_SOURCES "$<$<NOT:$<CONFIG:Debug>>:${simdjson_SOURCE_DIR}/cmake/empty.cpp>")
+  endif()
+
+  if(ARGS_COMPILE_ONLY)
+    list(APPEND ARGS_LABELS compile)
   endif()
 
   # Add the compile target
-  if (ARGS_LIBRARY)
+  if(ARGS_LIBRARY)
     add_library(${TEST_NAME} STATIC ${ARGS_SOURCES})
-  else(ARGS_LIBRARY)
+  else()
     add_executable(${TEST_NAME} ${ARGS_SOURCES})
-  endif(ARGS_LIBRARY)
+  endif()
 
   # Add test
-  if (ARGS_COMPILE_ONLY OR ARGS_LIBRARY)
+  if(ARGS_COMPILE_ONLY OR ARGS_LIBRARY)
```

Comment:
I take it no space after the if() is considered idiomatic for cmake?

Reply:
`if` is "just" a command like any other and the other commands aren't spaced apart either.

Reply:
I understand; just not sure what "most people" do in cmake :)